### ISSUE TITLE
use fp16 to train the t2t mT5 model

### ISSUE THREAD

**Issue by JUNSONG-112** [alexa/massive#26]
can we use half-precision training   similar as  --fp16 to fine tune the model?    as the t2t mT5 model is huge...
https://huggingface.co/transformers/v1.2.0/examples.html

**Comment by FeiWang96** [comment]
I've tried, but the model returns 'nan' loss. It seems to be a common issue of T5-style models: https://github.com/huggingface/transformers/issues/10819.

**Comment by jgmf-amazon** [comment]
Indeed, I've seen this problem with the T5-style models previously. I haven't specifically tried with MASSIVE. It's likely due to (1) the use of bfloat16, which has a much higher dynamic range during pretraining and, relatedly,  (2) the high variance in the initial random weights, which also isn't a problem for bfloat16. (This is discussed further [here](https://discuss.huggingface.co/t/mixed-precision-for-bfloat16-pretrained-models/5315) FYI).

**Comment by jgmf-amazon** [comment]
I probably don't have bandwidth to code this up for MASSIVE, but one option is to selectively convert to FP32 before operations that overflow. You can see where someone already implemented this as an example [here](https://github.com/huggingface/transformers/blob/v4.20.1/src/transformers/models/t5/modeling_t5.py#L255), but there may be other spots where it could be done.

**Comment by jgmf-amazon** [comment]
I'm marking this as an enhancement, namely to provide FP16 support for T5-based models.

**Comment by tomsherborne** [comment]
@jgmf-amazon @JUNSONG-112 
I've tested a model run using fp32, bf16 and tf32 as newer low-precision widths which don't have the fp16 issues. 

Using the config from #27 identically with an increased batch size to 1024 (assuming you have access to 1xA100) this is what I found:

`fp32` en-US validation:
```
  "eval_en-US_ex_match_acc": 0.6591796875,
  "eval_en-US_ex_match_acc_stderr": 0.014812032794827365,
  "eval_en-US_intent_acc": 0.8701171875,
  "eval_en-US_intent_acc_stderr": 0.010505451873177947,
  "eval_en-US_loss": 0.9552506804466248,
  "eval_en-US_runtime": 1.8402,
  "eval_en-US_samples_per_second": 1104.798,
  "eval_en-US_slot_micro_f1": 0.7760109041344843,
  "eval_en-US_slot_micro_f1_stderr": 0.002234384517660625,
  "eval_en-US_steps_per_second": 1.087,
```

`tf32` en-US validation:
```
  "eval_en-US_ex_match_acc": 0.6650390625,
  "eval_en-US_ex_match_acc_stderr": 0.014749275268514292,
  "eval_en-US_intent_acc": 0.873046875,
  "eval_en-US_intent_acc_stderr": 0.010403764204450758,
  "eval_en-US_loss": 0.9435577392578125,
  "eval_en-US_runtime": 1.39,
  "eval_en-US_samples_per_second": 1462.627,
  "eval_en-US_slot_micro_f1": 0.7781818181818182,
  "eval_en-US_slot_micro_f1_stderr": 0.002226638292266889,
  "eval_en-US_steps_per_second": 1.439,
```

`bf16` en-US validation:
```
"eval_en-US_ex_match_acc": 0.662109375,
  "eval_en-US_ex_match_acc_stderr": 0.014780970879272303,
  "eval_en-US_intent_acc": 0.8740234375,
  "eval_en-US_intent_acc_stderr": 0.010369467095865153,
  "eval_en-US_loss": 0.9545955657958984,
  "eval_en-US_runtime": 1.4358,
  "eval_en-US_samples_per_second": 1415.896,
  "eval_en-US_slot_micro_f1": 0.7760109041344843,
  "eval_en-US_slot_micro_f1_stderr": 0.002234384517660625,
  "eval_en-US_steps_per_second": 1.393,
```

Either approach approximately cuts the runtime in half (and more if you turn off all validation). If this looks reasonable I can PR a suggestion for `tf32` or `bf16`. I havent tested this on mT5 yet but will shortly.
